File: packages/cli/src/services/repository-ignore.test.ts
import { mkdir, readFile, rm, writeFile } from "node:fs/promises";
import { join } from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { createTemporaryDirectory } from "../test/helpers/sync-fixture.ts";
import {
  ensureManagedSecretArtifactIgnoreRules,
  managedSecretArtifactIgnoreBlock,
} from "./repository-ignore.ts";

const temporaryDirectories: string[] = [];

const createWorkspace = async () => {
  const directory = await createTemporaryDirectory("dotweave-ignore-test-");

  temporaryDirectories.push(directory);

  return directory;
};

afterEach(async () => {
  while (temporaryDirectories.length > 0) {
    const directory = temporaryDirectories.pop();

    if (directory !== undefined) {
      await rm(directory, { force: true, recursive: true });
    }
  }
});

describe("repository ignore rules", () => {
  it("creates the managed secret artifact block", async () => {
    const workspace = await createWorkspace();

    await ensureManagedSecretArtifactIgnoreRules(workspace);

    expect(await readFile(join(workspace, ".gitignore"), "utf8")).toBe(
      managedSecretArtifactIgnoreBlock,
    );
  });

  it("appends the managed block after existing user rules", async () => {
    const workspace = await createWorkspace();
    const ignorePath = join(workspace, ".gitignore");

    await mkdir(workspace, { recursive: true });
    await writeFile(ignorePath, "*.dotweave.secret\nbuild/\n", "utf8");

    await ensureManagedSecretArtifactIgnoreRules(workspace);

    expect(await readFile(ignorePath, "utf8")).toBe(
      `*.dotweave.secret\nbuild/\n${managedSecretArtifactIgnoreBlock}`,
    );
  });

  it("replaces an existing managed block without duplicating it", async () => {
    const workspace = await createWorkspace();
    const ignorePath = join(workspace, ".gitignore");

    await mkdir(workspace, { recursive: true });
    await writeFile(
      ignorePath,
      `node_modules/\n# BEGIN dotweave managed secret artifact rules\nold\n# END dotweave managed secret artifact rules\n*.log\n`,
      "utf8",
    );

    await ensureManagedSecretArtifactIgnoreRules(workspace);
    await ensureManagedSecretArtifactIgnoreRules(workspace);

    expect(await readFile(ignorePath, "utf8")).toBe(
      `node_modules/\n*.log\n${managedSecretArtifactIgnoreBlock}`,
    );
  });
});

File: packages/cli/test/services/repository_ignore_test.dart
import 'dart:io';

import 'package:dotweave/src/services/repository_ignore.dart';
import 'package:path/path.dart' as p;
import 'package:test/test.dart';

void main() {
  final temporaryDirectories = <String>[];

  Future<String> createWorkspace() async {
    final directory = await Directory.systemTemp.createTemp(
      'dotweave-ignore-test-',
    );

    temporaryDirectories.add(directory.path);

    return directory.path;
  }

  tearDown(() async {
    while (temporaryDirectories.isNotEmpty) {
      final directory = temporaryDirectories.removeLast();

      if (await Directory(directory).exists()) {
        await Directory(directory).delete(recursive: true);
      }
    }
  });

  group('repository ignore rules', () {
    test('creates the managed secret artifact block', () async {
      final workspace = await createWorkspace();

      await ensureManagedSecretArtifactIgnoreRules(workspace);

      expect(
        await File(p.join(workspace, '.gitignore')).readAsString(),
        managedSecretArtifactIgnoreBlock,
      );
    });

    test('appends the managed block after existing user rules', () async {
      final workspace = await createWorkspace();
      final ignorePath = p.join(workspace, '.gitignore');

      await Directory(workspace).create(recursive: true);
      await File(ignorePath).writeAsString('*.dotweave.secret\nbuild/\n');

      await ensureManagedSecretArtifactIgnoreRules(workspace);

      expect(
        await File(ignorePath).readAsString(),
        '*.dotweave.secret\nbuild/\n$managedSecretArtifactIgnoreBlock',
      );
    });

    test('replaces an existing managed block without duplicating it', () async {
      final workspace = await createWorkspace();
      final ignorePath = p.join(workspace, '.gitignore');

      await Directory(workspace).create(recursive: true);
      await File(ignorePath).writeAsString(
        'node_modules/\n'
        '# BEGIN dotweave managed secret artifact rules\n'
        'old\n'
        '# END dotweave managed secret artifact rules\n'
        '*.log\n',
      );

      await ensureManagedSecretArtifactIgnoreRules(workspace);
      await ensureManagedSecretArtifactIgnoreRules(workspace);

      expect(
        await File(ignorePath).readAsString(),
        'node_modules/\n*.log\n$managedSecretArtifactIgnoreBlock',
      );
    });
  });
}
